### PR TITLE
[Aptos Framework][Coin] Switch to an explicit opt-out model for coin transfers

### PR DESCRIPTION
**Pull request by movekevin**
### Description

Context:
Currently, in many cases, coins (ERC-20 tokens, including APT) cannot be sent directly to an account if:

- The account has not been created. Aptos_account::transfer or account::create_account needs to be called by another account that can pay for gas. This is generally a slight annoyance but not too big of a pain.
- The account has not registered to receive that coin type. This has to be done for every custom coin (APT is registered by default when creating an account). This is the main cause of complaints/pain.

Proposed solution:
Switch over to an explicit opt-out model. By default, an account can receive transfers of coins they have not registered via aptos_account::transfer_coins. They can explicitly opt-out any time via aptos_account:: set_allow_direct_coin_transfers

### Test Plan
Unit tests

### CODE REVIEW

File: aptos-move/framework/aptos-framework/sources/aptos_account.move [hunk 1]
```diff
@@ -12,6 +15,28 @@ module aptos_framework::aptos_account {
...
+    /// Configuration for whether an account can receive direct transfers of coins that they have not registered.
+    ///
+    /// By default, this is enabled. Users can opt-out by disabling at any time.
+    struct DirectTransferConfig has key {
+        allow_arbitrary_coin_transfers: bool,
+        update_coin_transfer_events: EventHandle<DirectCoinTransferConfigUpdated>,
+    }
+
+    /// Event emitted when an account's direct coins transfer config is updated.
+    struct DirectCoinTransferConfigUpdated has drop, store {
```
Comment (areshand) [inline aptos-move/framework/aptos-framework/sources/aptos_account.move:32]: do we want to call this DirectCoinTransferConfigUpdateEvent or sth to indicate this is an event?

File: aptos-move/framework/aptos-framework/sources/aptos_account.move [hunk 2]
```diff
@@ -22,13 +47,37 @@ module aptos_framework::aptos_account {
...
 
+    /// Convenient function to transfer a custom CoinType to a recipient account that might not exist.
+    /// This would create the recipient account first and register it to receive the CoinType, before transferring.
+    public entry fun transfer_coins<CoinType>(from: &signer, to: address, amount: u64) acquires DirectTransferConfig {
+        deposit_coins(to, coin::withdraw<CoinType>(from, amount));
+    }
+
+    /// Convenient function to deposit a custom CoinType into a recipient account that might not exist.
+    /// This would create the recipient account first and register it to receive the CoinType, before transferring.
+    public entry fun deposit_coins<CoinType>(to: address, coins: Coin<CoinType>) acquires DirectTransferConfig {
```
Comment (areshand) [inline aptos-move/framework/aptos-framework/sources/aptos_account.move:67]: do we want a name showing this is more than just depositing coins? For example, initialize_account_with_coin_deposit ? I am bad at naming... just throw some idea here

File: aptos-move/framework/aptos-framework/sources/aptos_account.move [hunk 3]
```diff
@@ -12,6 +15,28 @@ module aptos_framework::aptos_account {
...
+        update_coin_transfer_events: EventHandle<DirectCoinTransferConfigUpdated>,
+    }
+
+    /// Event emitted when an account's direct coins transfer config is updated.
+    struct DirectCoinTransferConfigUpdated has drop, store {
+        new_allow_direct_transfers: bool,
+    }
+
+    /// Event emitted when an account's direct NFT token transfer config is updated.
+    struct DirectTokenTransferConfigUpdated has drop, store {
```
Comment (areshand) [inline aptos-move/framework/aptos-framework/sources/aptos_account.move:37]: do we use this anywhere?

Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/aptos_account.move]: This is an artifact from my experiment earlier. Removed now.

File: aptos-move/framework/aptos-framework/sources/aptos_account.move (hunk 1 above)
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/aptos_account.move]: Updated

File: aptos-move/framework/aptos-framework/sources/aptos_account.move (hunk 2 above)
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/aptos_account.move:62]: I think deposit_coins is fine. This is meant to initialize behind the scene if necessary, similar to aptos_account::transfer, which can create the account if it doesn't exist behind the scene.

**Review by areshand — APPROVED** [review]

File: aptos-move/framework/aptos-framework/sources/aptos_account.move [hunk 4]
```diff
@@ -4,6 +4,9 @@ module aptos_framework::aptos_account {
     use aptos_framework::account;
     use aptos_framework::aptos_coin::AptosCoin;
     use aptos_framework::coin;
+    use aptos_framework::event::{EventHandle, emit_event};
+    use std::signer;
```
Comment (0xchloe) [inline aptos-move/framework/aptos-framework/sources/aptos_account.move:8]: nit: import not in order

File: aptos-move/framework/aptos-framework/sources/aptos_account.move [hunk 5]
```diff
@@ -22,13 +42,37 @@ module aptos_framework::aptos_account {
         coin::register<AptosCoin>(&signer);
     }
 
+    /// Convenient function to transfer APT to a recipient account that might not exist.
+    /// This would create the recipient account first, which also registers it to receive APT, before transferring.
     public entry fun transfer(source: &signer, to: address, amount: u64) {
```
Comment (0xchloe) [inline aptos-move/framework/aptos-framework/sources/aptos_account.move:47]: nit: should we call this function by a more explicit name? smth like `transfer_to_existing_or_nonexisting_account` - this does sound way more verbose than `transfer`, but think `transfer` is a simple name that people might mis-use

**Review by 0xchloe — APPROVED** [review]

File: aptos-move/framework/aptos-framework/sources/aptos_account.move (hunk 5 above)
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/aptos_account.move:47]: That's really verbose. I'm not sure how people would misuse transfer here as it's meant to implicitly create an account (and may be register the coin type) if one doesn't exist yet. As proposed (I'll create an AIP soon), this new flow/function should be the default to use so users wouldn't need to worry about the nuances of accounts and coin registration.

File: aptos-move/framework/aptos-framework/sources/aptos_account.move (hunk 4 above)
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/aptos_account.move:7]: Updated.

File: aptos-move/framework/aptos-framework/sources/aptos_account.move (hunk 5 above)
Comment (0xchloe) [reply aptos-move/framework/aptos-framework/sources/aptos_account.move:45]: yeah I don't feel strong about this, just thought that some devs might use the transfer function here as the default transfer function without realizing that they could be transferring coins to an account that doesn't exist yet. as long as we make it clear, sgtm

**Pull request merged**